### PR TITLE
Initial support for microchip cec1702

### PR DESCRIPTION
**Pull request by fabled**
@galak Finally got to making this PR as promised way earlier. Please take a look if things are done in a sane way, or need fixing.

### CODE REVIEW

**Review by galak — CHANGES_REQUESTED** [review]
The board, SoC, and HAL import should all be separate commits.  The HAL commit message should conform to what is specified in CONTRIBUTING.rst

**Review by galak — CHANGES_REQUESTED** [review]
All *.c & *.h files should have copyright/SPDX license info. 

Also missing board docs.  All new boards require board docs.

**Comment by fabled** [comment]
@galak Thanks, will fix accordingly.

Any comments on the Kconfig things? I was not fully sure how to name/use some of the board specifics, but I did what seemed to make sense and works.

`UART_NS16550_PORT_0_CLK_FREQ` is also hard coded in the fixup. Wonder how to put that into the dts.

I am next working on SPI and SPI Flash drivers. Is there any work on a generic SPI flash driver that I could reuse? I will need few flash specific overrides, but otherwise it's pretty generic chip on the demo board: SST26VF016B.

PINCTRL and GPIO drivers will probably come too later.

File: arch/arm/soc/microchip_cec/soc_config.c [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+static int soc_init(struct device *dev)
+{
+	__enable_irq();					/* Enable IRQs as ROM loader does PRIMASK=1 */
+	PCR->CLK_REQ_1 |= 1 << 0;			/* Clock: Interrupt block */
+	PCR->CLK_REQ_1 |= 1 << 8;			/* Clock: Processor */
+#ifdef CONFIG_UART_NS16550_PORT_0
+	PCR->CLK_REQ_2 |= 1 << 1;			/* Clock: UART0 */
+	GPIO_100_137->GPIO_104_PIN_CONTROL = 1 << 12;	/* GPIO104 Alt1 = UART0_TX */
+	GPIO_100_137->GPIO_105_PIN_CONTROL = 1 << 12;	/* GPIO105 Alt1 = UART0_RX */
```
Comment (MaureenHelm) [inline arch/arm/soc/microchip_cec/soc_config.c]: We usually do pinmuxing in the board init rather than the SoC init

File: boards/arm/secureiot1702/secureiot1702.dts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+		zephyr,sram = &sram0;
+		zephyr,console = &uart1;
+		zephyr,flash = &flash0;
+	};
+
+	/*
+	GPIO054 = LED Red
+	GPIO053 = LED Green
+	GPIO001 = LED Blue
+	GPIO157 = LED White
```
Comment (MaureenHelm) [inline boards/arm/secureiot1702/secureiot1702.dts]: This comment block should be removed

File: boards/arm/secureiot1702/secureiot1702_defconfig [hunk 3]
```diff
@@ -0,0 +1,17 @@
+CONFIG_ARM=y
+CONFIG_SOC_FAMILY_CEC=y
+CONFIG_SOC_CEC1702=y
+CONFIG_BOARD_SECUREIOT1702=y
+CONFIG_CORTEX_M_SYSTICK=y
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=48000000
+CONFIG_HAS_DTS=y
```
Comment (MaureenHelm) [inline boards/arm/secureiot1702/secureiot1702_defconfig]: This line shouldn't be here

File: ext/hal/Kconfig [hunk 4]
```diff
@@ -14,6 +14,8 @@ source "ext/hal/atmel/asf/Kconfig"
 
 source "ext/hal/altera/Kconfig"
 
+source "ext/hal/cec/Kconfig"
```
Comment (MaureenHelm) [inline ext/hal/Kconfig]: Let's use "ext/hal/microchip/cec" instead

File: ext/hal/cec/include/system_MEC2016.h [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+* PRODUCTS, COMBINATION WITH ANY OTHER PRODUCTS, OR USE IN ANY APPLICATION.
+* IN NO EVENT WILL MICROCHIP BE LIABLE FOR ANY INDIRECT, SPECIAL, PUNITIVE,
+* INCIDENTAL OR CONSEQUENTIAL LOSS, DAMAGE, COST OR EXPENSE OF ANY KIND
+* WHATSOEVER RELATED TO THE SOFTWARE, HOWEVER CAUSED, EVEN IF MICROCHIP HAS
+* BEEN ADVISED OF THE POSSIBILITY OR THE DAMAGES ARE FORESEEABLE.
+* TO THE FULLEST EXTENT ALLOWED BY LAW, MICROCHIP'S TOTAL LIABILITY ON ALL
+* CLAIMS IN ANY WAY RELATED TO THIS SOFTWARE WILL NOT EXCEED THE AMOUNT OF
+* FEES, IF ANY, THAT YOU HAVE PAID DIRECTLY TO MICROCHIP FOR THIS SOFTWARE.
+* MICROCHIP PROVIDES THIS SOFTWARE CONDITIONALLY UPON YOUR ACCEPTANCE
+* OF THESE TERMS.
```
Comment (MaureenHelm) [inline ext/hal/cec/include/system_MEC2016.h]: This looks like a new license and will need approval by the gov board

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: arch/arm/soc/microchip_cec/soc_config.c (hunk 1 above)
Comment (fabled) [reply arch/arm/soc/microchip_cec/soc_config.c]: Ah, ok. I'll move the board specifics there. Eventually the pinmuxing will be done from the dts when pinctrl driver works.

File: ext/hal/cec/include/system_MEC2016.h (hunk 5 above)
Comment (fabled) [reply ext/hal/cec/include/system_MEC2016.h]: This file should not be there. I'll remove it, it's unneeded.

File: ext/hal/Kconfig (hunk 4 above)
Comment (fabled) [reply ext/hal/Kconfig]: I make it cec1702 then. There's also cec1302 which is older version. Has same cpu but different peripheral blocks (especially the crypto side).

**Comment by fabled** [comment]
Change requests addressed. Please review again. Copyright notices also added. This work from @CryptaLabs.

File: arch/arm/soc/microchip_cec/Kconfig.soc [hunk 6]
```diff
@@ -0,0 +1,9 @@
+config SOC_SERIES_CEC
+	bool "Microchip CECxxxx MCU"
+	select CPU_CORTEX_M
+	select CPU_CORTEX_M4
+	select CPU_FAMILY_CEC
```
Comment (dbkinder) [inline arch/arm/soc/microchip_cec/Kconfig.soc]: CPU_FAMILY_CEC seems undefined

File: boards/arm/secureiot1702/Kconfig.board [hunk 7]
```diff
@@ -0,0 +1,5 @@
+config BOARD_SECUREIOT1702
+	bool "Microchip SecureIoT1702"
+	depends on SOC_SERIES_CEC
+	select HAS_DTS
+	select SOC_PART_NUMBER_CEC1702
```
Comment (dbkinder) [inline boards/arm/secureiot1702/Kconfig.board]: SOC_PART_NUMBER_CEC1702 seems undefined

File: boards/arm/secureiot1702/doc/board.rst [hunk 8]
```diff
@@ -0,0 +1,18 @@
+.. _secureiot1702:
+
+Microchip SecureIoT1702
+#######################
+
+Board configuration for Microchip CEC1702 demo board. This is
```
Comment (dbkinder) [inline boards/arm/secureiot1702/doc/board.rst]: Thanks for getting the documentation started. We've got a recommended template for board documentation at doc/templates/board.tmpl (in your cloned copy of the zephyr tree).

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: arch/arm/soc/microchip_cec/Kconfig.soc (hunk 6 above)
Comment (fabled) [reply arch/arm/soc/microchip_cec/Kconfig.soc]: Right. This was little bit hazy for me on how and where to define `SOC_FAMILY`, `SOC_SERIES` and which type of `CPU_{FAMILY,SERIES}_CEC` things to define. Having that config split to `Kconfig`, `Kconfig.soc`, `Kconfig.defconfig`, etc. could be little bit less cluttered. I ended up not defining the family at all since there seems to be only two CEC CPUs at the moment. Maybe I'll just change this to `CPU_SERIES_CEC`

File: boards/arm/secureiot1702/Kconfig.board (hunk 7 above)
Comment (fabled) [reply boards/arm/secureiot1702/Kconfig.board]: Same here. I just change this to `SOC_CEC1702`.

File: boards/arm/secureiot1702/doc/board.rst (hunk 8 above)
Comment (fabled) [reply boards/arm/secureiot1702/doc/board.rst]: Ok, will take it from there. I also have pending patches for IRQ support, and QMSPI host driver for CEC. As well as generic SPI flash (auto-detects way to access chip using SFDP) driver. I will possibly also work on gpio, pinmux, clock drivers.

**Comment by galak** [comment]
Can update this to current and I'll review again towards getting it in tree

**Comment by fabled** [comment]
Rebase against master, addressed most checkpatch issues.

**Comment by galak** [comment]
@dbkinder are the docs good now?

File: boards/arm/secureiot1702/doc/secureiot1702.rst [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+#######################
+
+Overview
+********
+
+Board configuration for Microchip CEC1702 demo board. This is a cryptographic
+embedded controlled based on ARM Cortex-M4.
+
+Highlights of the board:
+- CEC1702 32-bit ARM® Cortex®-M4F Controller with Integrated Crypto
```
Comment (dbkinder) [inline boards/arm/secureiot1702/doc/secureiot1702.rst]: need a blank line before a bullet list

File: boards/arm/secureiot1702/doc/secureiot1702.rst [hunk 10]
```diff
@@ -0,0 +1,66 @@
+.. _secureiot1702:
+
+Microchip SecureIoT1702
+#######################
+
+Overview
+********
+
+Board configuration for Microchip CEC1702 demo board. This is a cryptographic
+embedded controlled based on ARM Cortex-M4.
```
Comment (dbkinder) [inline boards/arm/secureiot1702/doc/secureiot1702.rst]: on an ARM

File: boards/arm/secureiot1702/doc/secureiot1702.rst [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+- Serial Quad I/O (SQI) flash
+- OTP programmable in CEC1702
+- Two expansion headers compatible with MikroElektronika mikroBUS™ Expansion interface
+
+
+Supported Features
+==================
+
+Following devices are supported:
+- Nested Vectored Interrupt Controller (NVIC)
```
Comment (dbkinder) [inline boards/arm/secureiot1702/doc/secureiot1702.rst]: need a blank line before a bullet list

File: boards/arm/secureiot1702/doc/secureiot1702.rst [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+Please refer to the demo board schematics for the pin routings. Additional
+devices can be connected via mikroBUS expansion interface.
+
+Programming and Debugging
+*************************
+
+Flashing
+========
+[How to use this board with Zephyr and how to flash a Zephyr binary on this
```
Comment (dbkinder) [inline boards/arm/secureiot1702/doc/secureiot1702.rst]: Information missing (did you intend to publish without this filled out)?

File: boards/arm/secureiot1702/doc/secureiot1702.rst [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+device]
+
+
+Debugging
+=========
+[ How to debug this board]
```
Comment (dbkinder) [inline boards/arm/secureiot1702/doc/secureiot1702.rst]: Information missing

File: boards/arm/secureiot1702/doc/secureiot1702.rst [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+
+References
+**********
+
```
Comment (dbkinder) [inline boards/arm/secureiot1702/doc/secureiot1702.rst]: These target link definitions won't show up here in the generated output.  If you'd like them to appear, you can add a directive here:

    .. target_notes::
         
(be sure there's a blank line after this directive).  If you don't want these links to show up here in the output, then delete the ``References`` heading above (and don't put in the target_notes directive :)

**Review by dbkinder — CHANGES_REQUESTED** [review]
Some formatting issues and there are sections with no content.

**Comment by fabled** [comment]
Trivial documentation issues fixed. The flashing instructions are omitted at this time because additional software is required to modify the produced `.bin` image to bootable image. I have rough python code working to do it, but it needs clean up and integration to the cmake system. I have not used any debugger yet with the CPU, though it does have a standard JTAG support in it.

**Comment by galak** [comment]
Sorry for not saying this earlier, but wondering if we can either remove the HAL code import, or if we keep it we since its not Apache licensed we need to go through the process documented here:

http://docs.zephyrproject.org/contribute/contribute_non-apache.html

**Review by dbkinder — CHANGES_REQUESTED** [review]
Flashing and Debugging sections are still empty

**Comment by fabled** [comment]
@galak 
> Sorry for not saying this earlier, but wondering if we can either remove the HAL code import, or if we keep it we since its not Apache licensed we need to go through the process documented here:
> 
> http://docs.zephyrproject.org/contribute/contribute_non-apache.html

The HAL is used in the board pin mux setup. It may be also needed for few other places. Of course much of the code can be written be rewriting the relevant things on per-driver basis.

There is additional ROM API present. It includes some header files and symbols introduced in the linker script. I have pending commit for that too, and may need it in.

@dbkinder 
> Flashing and Debugging sections are still empty

The CEC1702 does not contain flash. It requires external SPI flash chip that one needs to program using externally via the SPI programmer. This is true for the demo board, it's flashed by connecting directly to the SPI bus. I personally used rpi + flashrom, but other use cases exist. I wonder how to properly document this.

There is standard JTAG option available, but I have not used it. Apparently it needs also additional steps on the MCU to enable it, as they intentionally disable debugging for ROM boot up code.

Basically the chip boots from ROM, that loads the main firmware image from the external SPI chip. The image may be encrypted and signed. There's eFuse bits that contain the encryption/signature keys and the information if signature is required or not.

Lot of this is outside the scope of the initial commits. So I wonder if we drop the section, or briefly explain the above?

**Comment by galak** [comment]
@fabled still interested in getting this merged?

**Comment by fabled** [comment]
@galak Yes, would be nice. Seems a rebase is needed... I can do that in near future. Is there other changes that would still be needed?

**Comment by fabled** [comment]
@galak Rebased, though testing is still in progress.

**Review by franciscomunoz — CHANGES_REQUESTED** [review]

**Comment by fabled** [comment]
@franciscomunoz Requested changes added.
Fixed to advertise FPU support, and reworked the config related matters so things compile again. QMSPI code style issues fixed.

@franciscomunoz @galak Anything else to fix?

**Comment by franciscomunoz** [comment]
Timo,

It would be good if you rename the file  spi_cec_qmspi.c to spi_xec_qmspi.c
. This is because the whole Microchip family is MEC/CEC. What do you think
?

—Francisco

**Comment by fabled** [comment]
Then we would need to rename a lot of other things too. CEC is assumed as the SoC family in many places.

The MEC170x seems to be new set of chips that are CEC compatible. The docs seem to speak of "CEC/MEC family". However, there's a lot of other MEC series chips that are not compatible. There's also CEC1302 which is earlier revision that seems more or less compatible.

The headers seem to speak of MEC2016 as the core. So I'd probably keep the 'CEC' name. The MEC170x subset is just CEC compatible. Alternatively, if going for MEC, we should use MEC2016 or MEC170x as there's bunch of other MEC chips that are completely different. Let's stay with CEC?

**Comment by franciscomunoz** [comment]
I see, sure no problem.

**Comment by fabled** [comment]
@galak Ping.

**Comment by galak** [comment]
> @galak Ping.

Its been a while, sorry about that.  Since this out of my memory, my first question is what is the license of the CEC HAL code?

**Comment by fabled** [comment]
It seems to be pretty free. The header file carries this:
```
 * @par      ARM Limited (ARM) is supplying this software for use with Cortex-M
 *           processor based microcontroller, but can be equally used for other
 *           suitable processor architectures. This file can be freely distributed.
 *           Modifications to this file shall be clearly marked.
 *           
 *           THIS SOFTWARE IS PROVIDED "AS IS". NO WARRANTIES, WHETHER EXPRESS, IMPLIED
 *           OR STATUTORY, INCLUDING, BUT NOT LIMITED TO, IMPLIED WARRANTIES OF
 *           MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE APPLY TO THIS SOFTWARE.
 *           ARM SHALL NOT, IN ANY CIRCUMSTANCES, BE LIABLE FOR SPECIAL, INCIDENTAL, OR
 *           CONSEQUENTIAL DAMAGES, FOR ANY REASON WHATSOEVER. 
```

**Comment by galak** [comment]
Do you mind bring this PR up to date w/latest mater.  I need to start a license review for the HAL since its not using a standard license.  I don't see any issues w/it.

**Comment by fabled** [comment]
@galak The conflict is simple to resolve. But seems the SoC stuff resides in new directory. Could you point an example commit how to add a new soc / soc-family (+ soc-series if needed) on the new layout and what the Kconfig files should be named like?

**Comment by galak** [comment]
> @galak The conflict is simple to resolve. But seems the SoC stuff resides in new directory. Could you point an example commit how to add a new soc / soc-family (+ soc-series if needed) on the new layout and what the Kconfig files should be named like?

Sure, pretty much we just moved `arch/arm/soc/microchip_cec` to `soc/arm/microchip_cec`.  Take a look at:

commit 537798db0d82c1b6c6f2d6fb029210ef007ebf6e
Author: Diego Sueiro <diego.sueiro@gmail.com>
Date:   Wed Oct 17 07:58:14 2018 +0100

    soc: arm: exx32: Add Silabs EFR32MG12P soc files

**Comment by fabled** [comment]
@galak Mechanical rebase done. Builds again the hello sample. Unfortunately, have not boot tested yet.

**Comment by galak** [comment]
> @galak Mechanical rebase done. Builds again the hello sample. Unfortunately, have not boot tested yet.

thanks, I submitted the license for review, with the Thanksgiving Holiday in the US its going to delay things a bit.

**Review by galak — CHANGES_REQUESTED** [review]
So the whole CONFIG_UART_NS16550_PORT_0_BRD_MASK needs to be changed.

1. I don't see this used anywhere
2. If we need this, do it for the more specific feature you need (ie alternate clock source)

File: boards/arm/secureiot1702/board.h [hunk 15]
```diff
@@ -0,0 +1,6 @@
+#ifndef __INC_BOARD_H
```
Comment (galak) [inline boards/arm/secureiot1702/board.h]: remove board.h

File: boards/arm/secureiot1702/dts_fixup.h [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+ * are modified to handle the generated information, or the mapping of
+ * generated data matches the driver definitions.
+ */
+
+
+#define DT_NUM_IRQ_PRIO_BITS			DT_ARM_V7M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
+
+#define CONFIG_NS16550_REG_SHIFT		DT_NS16550_400F2400_REG_SHIFT
+#define DT_UART_NS16550_PORT_0_BASE_ADDR	DT_NS16550_400F2400_BASE_ADDRESS
+#define DT_UART_NS16550_PORT_0_CLK_FREQ		1843200
```
Comment (galak) [inline boards/arm/secureiot1702/dts_fixup.h]: clock frequency should be a property in the .dts file

**Review by galak — CHANGES_REQUESTED** [review]
Some updates to the board code needed.

**Review by galak — CHANGES_REQUESTED** [review]
The qmspi change seems to be missing a dts binding file.  Also I'd like to pull that PR out so its get seen by the SPI guys.

**Comment by galak** [comment]
@tbursztyka Can you look at the SPI driver in this PR.

**Comment by fabled** [comment]
@galak Ping. Reworked The NS16550 UART alternate clock select to have SoC specific block in the uart driver. That simplifies things quite a bit. Let me know if would have a better approach. I hope we could get this merged in soonish. It's been pending a long time...

**Review by dbkinder — CHANGES_REQUESTED** [review]
Documentation is better, but still missing Flashing and Debugging section content.

**Comment by dbkinder** [comment]
@nashif @galak What should we do with PRs that are a year old and still open?

**Comment by fabled** [comment]
@dbkinder 

> Documentation is better, but still missing Flashing and Debugging section content.

I will work out the Flashing section. The vendor provider stuff is all Windows -based. Getting it done in generic way is doable, but may need some extras. Any way it boots from external SPI chip, so the thing is to have SPI flasher. What comes to Debugging, I don't have much experience as I never used hardware debugger. The SoC and board has standard JTAG support, but I basically never used it. So It is not really possible for me anything else that "Standard JTAG supported." or similar. Would that be sufficient for the time being?

> What should we do with PRs that are a year old and still open?

Much of the time has been waiting review - or scheduling some cycles to this. Howevre, I am still happy to do requested changes as possible. I really hope we could get this to acceptable state so others can start building on it. I have had some private questions asked on this. There is now another demo board built on this SoC, and there is also four new SoCs (MEC1701, MEC1703, MEC1704 and MEC1705) which are of same core/family.

**Comment by nashif** [comment]
closing as stale
